### ISSUE TITLE
Occasional ClassCastExceptions occurring with conditionally shown views

### ISSUE THREAD

**Issue by christopherniksch** [airbnb/epoxy#1041]
Epoxy Version - `3.11.0`

We've noticed a handful of `ClassCastException`s occurring. Over the course of a day with at least a thousand visitors to a screen, we're seeing one, maybe two crashes on our fragment. The crashes all look like the following stacktrace.

`Fatal Exception: java.lang.ClassCastException`
```
com.x.y.z.ItemDivider_.handlePreBind (ItemDivider_.java:22)
com.airbnb.epoxy.EpoxyViewHolder.bind (EpoxyViewHolder.java:53)
com.airbnb.epoxy.BaseEpoxyAdapter.onBindViewHolder (BaseEpoxyAdapter.java:104)
com.airbnb.epoxy.BaseEpoxyAdapter.onBindViewHolder (BaseEpoxyAdapter.java:19)
androidx.recyclerview.widget.RecyclerView$Adapter.bindViewHolder (RecyclerView.java:7107)
androidx.recyclerview.widget.RecyclerView$Recycler.tryBindViewHolderByDeadline (RecyclerView.java:6012)
androidx.recyclerview.widget.RecyclerView$Recycler.tryGetViewHolderForPositionByDeadline (RecyclerView.java:6279)
androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition (RecyclerView.java:6118)
androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition (RecyclerView.java:6114)
androidx.recyclerview.widget.LinearLayoutManager$LayoutState.next (LinearLayoutManager.java:2303)
androidx.recyclerview.widget.LinearLayoutManager.layoutChunk (LinearLayoutManager.java:1627)
androidx.recyclerview.widget.LinearLayoutManager.fill (LinearLayoutManager.java:1587)
androidx.recyclerview.widget.LinearLayoutManager.onLayoutChildren (LinearLayoutManager.java:665)
androidx.recyclerview.widget.RecyclerView.dispatchLayoutStep2 (RecyclerView.java:4134)
androidx.recyclerview.widget.RecyclerView.onMeasure (RecyclerView.java:3540)
```

The EpoxyModel in question
```
@EpoxyModelClass(layout = R.layout.view_divider)
abstract class ItemDivider : EpoxyModelWithHolder<ItemDivider.DividerHolder>() {
  class DividerHolder : KotlinEpoxyHolder()
}
```

Its layout 
```
<?xml version="1.0" encoding="utf-8"?>
<View xmlns:android="http://schemas.android.com/apk/res/android"
  android:layout_width="match_parent"
  android:layout_height="@dimen/dp_1"
  android:background="@drawable/linear_layout_divider_vertical" />
```

The crashes are happening so far are occurring on four different views on the same screen. The other four views also vary in their complexity (non-empty holders, layouts containing a variety of child elements, etc). The only commonality between the views that causing these fatal exceptions is that they are all conditionally shown in the `EpoxyRecyclerView`.

For example, the following is a trimmed down version of what our code looks like. The state parameter to the function in this case is a data class contained a variety of fields which are the substates of the screen which get rendered out. The state comes from an rx stream in a `ViewModel`, and the substates all result from various network calls which update and emit a new state, with one relevant piece of the total state changed. Some substates are data classes, others are sealed classes. The `ClassCastException`s indicate that Models which are being conditionally shown, like the `ItemDivider` model, is trying to be cast to a view that immediately follows / precedes it in the list.
```
private fun buildLayout(state: MyScreenViewState) {
binding.epoxyRecyclerView.withModels {
      if (state.showProgress) {
        progressView { id("PROGRESS_INDICATOR") }
      } else {
        when (state.userState) {
          UserState.Hidden, is UserState.Visible -> {
            welcomeHeaderView {
              id("WELCOME_HEADER")
              userState(state.userState)
            }
          }
          UserState.Anonymous -> {
            signInView {
              id("SIGN_IN_VIEW")
            }
          }
        }

        if (state.BannerState is BannerState.Visible) {
          announcementBannerView {
            id("ANNOUNCEMENT_BANNER")
            viewState(state.BannerState)
        }

        if (state.userDetailState !is UserDetailState.Hidden) {
          itemDivider { id("DETAIL_TOP_DIVIDER") }
        }

        if (state.userState is UserState.Visible) {
          userView {
            id("USER_VIEW")
            userState(state.userState)
          }
          leftInsetDivider { id("DETAIL_BOTTOM_DIVIDER") }
        }

        spacer { id("SPACER_TWO") }

        if (state.userState is UserState.Visible) {
          paymentDetailsView {
            id("PAYMENT_DETAILS")
          }
          spacer { id("SPACER_FOUR") }
        }

      }
    }
  }

```

One final thing we've noticed from logs on BugSnag, is that crashes tend to happen when the fragment and activity its in pause & stop or start & resume. Also, I've been unable to reproduce these crashes locally when following user breadcrumbs to replicate what they were doing when the crash occurred. We've used epoxy extensively within our app without any similar crashes. Any thoughts to what could be going wrong here?

**Comment by elihart** [comment]
Thanks for the detailed report, can you share the full error message of the `ClassCastException` so I can see what the expected and actual types are? it looks like the pre bind method would be expecting a `DividerHolder`, but what is it actually receiving?

What it looks like is happening is that the underlying `EpoxyViewHolder` is being shared with another, different View. The view holders can be shared between any model with the same recyclerview view type, and the view type by default is taken from the layout. So if any other model uses the same layout `R.layout.view_divider` but has a different view holder class (ie not `ItemDivider. DividerHolder`) then this could happen

**Comment by christopherniksch** [comment]
The error message shows something along the lines of `com.x.y.z.ItemDividerHolder cannot be cast to com.x.y.z.UserView.ViewHolder`. Of note, we've had a couple more crashes pop up here and there from this screen indicating occasional other crashes are happening, but involving other `EpoxyModel`s. I've noticed the `ViewHolder`s involved in the crashes all correspond to models that are adjacent models in the `EpoxyRecyclerView`.

I have checked against what yourself and mark discussed on #854 and have found that every `EpoxyModel` used on this screen has its own specific `ViewHolder`, and every model has a distinct layout.

I did notice one thing that got left in when I migrated this screen to Epoxy. We originally were using the same `MyScreenViewState` to build the old layout and would use the substates to toggle the visibility of parts of the layout via KTX synthetics. While migrating the views over, one or two views were being given a substate of the `MyScreenViewState` that indicated they should be invisible, and were toggling their roots view's visibility to `View.GONE` while always being presented in the controller. Could an `EpoxyModel` added to the controller that internally makes itself hidden, cause a crash when the fragment containing this controller resumes?

**Comment by elihart** [comment]
That last bit about toggling visibility is interesting. I don't know for sure that could cause a crash, but it wouldn't surprise me if it was leading to unexpected behavior with RecyclerView. You may want to look through the RecyclerView source to see if anything stands out that looks like it could be affected, or at the very least discontinue that behavior.

I would also recommend updating to the latest recyclerview version if you haven't yet, as it sounds like the problem could be a bug internal to RecyclerView. I'm not sure what else to do without a clearer reproduction case :(

**Comment by christopherniksch** [comment]
Thanks for the suggestions. I'll delve into RecyclerView further to see if i can pinpoint anything. I've yet to reproduce this locally unfortunately. I have modified our screen to lift the visibility toggling logic out of the EpoxyModels and into the the Controller. I'll post a followup when we update our app if that change fixes things or not.

**Comment by christopherniksch** [comment]
I'm going to (finally) close this issue out, as it seems to be fixed now. While I was never able to reproduce this locally, it appears that the cause of the crash was from two of our `EpoxyModel`s toggling their visibility. For some context, we had some `ViewPager`s that would show cards to our users, typically 1 to 3 cards at a time. Based on some logging I collected of the screen state at the time, it appeared that the state of the cards was changing right as the crash occurred. The `ViewPager`s are wrapped into their own components which manage the appearance of the `ViewPager` differently when there's either a single card, or multiple. It appears that these card pagers were setting their root visibility to gone when all of the cards were dismissed by the user.

The solution I used (since we weren't rebuilding these card pagers in epoxy) was to wrap their layout in a `LinearLayout` so that the `EpoxtModel` wasn't technically being set to `GONE` and I also put in some better state handling code to prevent EpoxyModels from being added to the `EpoxyRecyclerView` if no cards were being shown. These two things combined, seem to have resolved this crash we were seeing.
